Index: sagemaker-code-editor/vscode/src/vs/platform/update/common/update.config.contribution.ts
===================================================================
--- sagemaker-code-editor.orig/vscode/src/vs/platform/update/common/update.config.contribution.ts
+++ sagemaker-code-editor/vscode/src/vs/platform/update/common/update.config.contribution.ts
@@ -50,7 +50,7 @@ configurationRegistry.registerConfigurat
 		},
 		'update.showReleaseNotes': {
 			type: 'boolean',
-			default: true,
+			default: false,
 			scope: ConfigurationScope.APPLICATION,
 			description: localize('showReleaseNotes', "Show Release Notes after an update. The Release Notes are fetched from a Microsoft online service."),
 			tags: ['usesOnlineServices']
