Index: sagemaker-code-editor/vscode/src/vs/platform/environment/common/environmentService.ts
===================================================================
--- sagemaker-code-editor.orig/vscode/src/vs/platform/environment/common/environmentService.ts
+++ sagemaker-code-editor/vscode/src/vs/platform/environment/common/environmentService.ts
@@ -101,7 +101,7 @@ export abstract class AbstractNativeEnvi
 			return URI.file(join(vscodePortable, 'argv.json'));
 		}
 
-		return joinPath(this.userHome, this.productService.dataFolderName, 'argv.json');
+		return joinPath(this.appSettingsHome, 'argv.json');
 	}
 
 	@memoize

Index: sagemaker-code-editor/vscode/src/vs/platform/languagePacks/browser/languagePacks.ts
===================================================================
--- sagemaker-code-editor.orig/vscode/src/vs/platform/languagePacks/browser/languagePacks.ts
+++ sagemaker-code-editor/vscode/src/vs/platform/languagePacks/browser/languagePacks.ts
@@ -5,18 +5,24 @@
 
 import { CancellationTokenSource } from '../../../base/common/cancellation.js';
 import { URI } from '../../../base/common/uri.js';
+import { ProxyChannel } from '../../../base/parts/ipc/common/ipc.js';
 import { IExtensionGalleryService } from '../../extensionManagement/common/extensionManagement.js';
 import { IExtensionResourceLoaderService } from '../../extensionResourceLoader/common/extensionResourceLoader.js';
-import { ILanguagePackItem, LanguagePackBaseService } from '../common/languagePacks.js';
+import { ILanguagePackItem, ILanguagePackService, LanguagePackBaseService } from '../common/languagePacks.js';
 import { ILogService } from '../../log/common/log.js';
+import { IRemoteAgentService } from '../../../workbench/services/remote/common/remoteAgentService.js';
 
 export class WebLanguagePacksService extends LanguagePackBaseService {
+	private readonly languagePackService: ILanguagePackService;
+
 	constructor(
+		@IRemoteAgentService remoteAgentService: IRemoteAgentService,
 		@IExtensionResourceLoaderService private readonly extensionResourceLoaderService: IExtensionResourceLoaderService,
 		@IExtensionGalleryService extensionGalleryService: IExtensionGalleryService,
 		@ILogService private readonly logService: ILogService
 	) {
 		super(extensionGalleryService);
+		this.languagePackService = ProxyChannel.toService<ILanguagePackService>(remoteAgentService.getConnection()!.getChannel('languagePacks'))
 	}
 
 	async getBuiltInExtensionTranslationsUri(id: string, language: string): Promise<URI | undefined> {
@@ -72,6 +78,6 @@ export class WebLanguagePacksService ext
 
 	// Web doesn't have a concept of language packs, so we just return an empty array
 	getInstalledLanguages(): Promise<ILanguagePackItem[]> {
-		return Promise.resolve([]);
+		return this.languagePackService.getInstalledLanguages()
 	}
 }
Index: sagemaker-code-editor/vscode/src/vs/server/node/remoteLanguagePacks.ts
===================================================================
--- sagemaker-code-editor.orig/vscode/src/vs/server/node/remoteLanguagePacks.ts
+++ sagemaker-code-editor/vscode/src/vs/server/node/remoteLanguagePacks.ts
@@ -3,37 +3,111 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/

+import * as path from 'path';
+import { promises as fs } from 'fs';
 import { FileAccess } from '../../base/common/network.js';
 import { join } from '../../base/common/path.js';
 import type { INLSConfiguration } from '../../nls.js';
 import { resolveNLSConfiguration } from '../../base/node/nls.js';
-import { Promises } from '../../base/node/pfs.js';
-import product from '../../platform/product/common/product.js';
 
 const nlsMetadataPath = join(FileAccess.asFileUri('').fsPath);
-const defaultMessagesFile = join(nlsMetadataPath, 'nls.messages.json');
 const nlsConfigurationCache = new Map<string, Promise<INLSConfiguration>>();
 
 export async function getNLSConfiguration(language: string, userDataPath: string): Promise<INLSConfiguration> {
-	if (!product.commit || !(await Promises.exists(defaultMessagesFile))) {
-		return {
-			userLocale: 'en',
-			osLocale: 'en',
-			resolvedLanguage: 'en',
-			defaultMessagesFile,
-
-			// NLS: below 2 are a relic from old times only used by vscode-nls and deprecated
-			locale: 'en',
-			availableLanguages: {}
-		};
-	}
-
 	const cacheKey = `${language}||${userDataPath}`;
 	let result = nlsConfigurationCache.get(cacheKey);
 	if (!result) {
-		result = resolveNLSConfiguration({ userLocale: language, osLocale: language, commit: product.commit, userDataPath, nlsMetadataPath });
+		// passing a dummy commit which is required to resolve language packs
+		result = resolveNLSConfiguration({ userLocale: language, osLocale: language, commit: 'dummy_commit', userDataPath, nlsMetadataPath });
 		nlsConfigurationCache.set(cacheKey, result);
+		// If the language pack does not yet exist, it defaults to English, which is
+		// then cached and you have to restart even if you then install the pack.
+		result.then((r) => {
+			if (!language.startsWith('en') && r.resolvedLanguage.startsWith('en')) {
+				nlsConfigurationCache.delete(cacheKey);
+			}
+		})
 	}
 
 	return result;
 }
+
+/**
+ * Copied from from src/main.js.
+ */
+export const getLocaleFromConfig = async (argvResource: string): Promise<string> => {
+	try {
+		const content = stripComments(await fs.readFile(argvResource, 'utf8'));
+		return JSON.parse(content).locale;
+	} catch (error) {
+		if (error.code !== "ENOENT") {
+			console.warn(error)
+		}
+		return 'en';
+	}
+};
+
+/**
+ * Copied from from src/main.js.
+ */
+const stripComments = (content: string): string => {
+	const regexp = /('(?:[^\\']*(?:\\.)?)*')|('(?:[^\\']*(?:\\.)?)*')|(\/\*(?:\r?\n|.)*?\*\/)|(\/{2,}.*?(?:(?:\r?\n)|$))/g;
+
+	return content.replace(regexp, (match, _m1, _m2, m3, m4) => {
+		// Only one of m1, m2, m3, m4 matches
+		if (m3) {
+			// A block comment. Replace with nothing
+			return '';
+		} else if (m4) {
+			// A line comment. If it ends in \r?\n then keep it.
+			const length_1 = m4.length;
+			if (length_1 > 2 && m4[length_1 - 1] === '\n') {
+				return m4[length_1 - 2] === '\r' ? '\r\n' : '\n';
+			}
+			else {
+				return '';
+			}
+		} else {
+			// We match a string
+			return match;
+		}
+	});
+};
+
+/**
+ * Generate translations then return a path to a JavaScript file that sets the
+ * translations into global variables.  This file is loaded by the browser to
+ * set global variables that the loader uses when looking for translations.
+ *
+ * Normally, VS Code pulls these files from a CDN but we want them to be local.
+ */
+export async function getBrowserNLSConfiguration(locale: string, userDataPath: string): Promise<string> {
+	if (locale.startsWith('en')) {
+		return ''; // Use fallback translations.
+	}
+
+	const nlsConfig = await getNLSConfiguration(locale, userDataPath);
+	const messagesFile = nlsConfig?.languagePack?.messagesFile;
+	const resolvedLanguage = nlsConfig?.resolvedLanguage;
+	if (!messagesFile || !resolvedLanguage) {
+		return ''; // Use fallback translations.
+	}
+
+	const nlsFile = path.join(path.dirname(messagesFile), "nls.messages.js");
+	try {
+		await fs.stat(nlsFile);
+		return nlsFile; // We already generated the file.
+	} catch (error) {
+		// ENOENT is fine, that just means we need to generate the file.
+		if (error.code !== 'ENOENT') {
+			throw error;
+		}
+	}
+
+	const messages = (await fs.readFile(messagesFile)).toString();
+	const content = `globalThis._VSCODE_NLS_MESSAGES=${messages};
+globalThis._VSCODE_NLS_LANGUAGE=${JSON.stringify(resolvedLanguage)};`
+	await fs.writeFile(nlsFile, content, "utf-8");
+
+	return nlsFile;
+}

Index: sagemaker-code-editor/vscode/src/vs/server/node/serverEnvironmentService.ts
===================================================================
--- sagemaker-code-editor.orig/vscode/src/vs/server/node/serverEnvironmentService.ts
+++ sagemaker-code-editor/vscode/src/vs/server/node/serverEnvironmentService.ts
@@ -14,6 +14,8 @@ import { URI } from 'vs/base/common/uri'
 
 export const serverOptions: OptionDescriptions<Required<ServerParsedArgs>> = {
 
+	'locale': { type: 'string' },
+
 	/* ----- server setup ----- */
 
 	'host': { type: 'string', cat: 'o', args: 'ip-address', description: nls.localize('host', "The host name or IP address the server should listen to. If not set, defaults to 'localhost'.") },
@@ -97,6 +99,8 @@ export const serverOptions: OptionDescri
 
 export interface ServerParsedArgs {
 
+	'locale'?: string;
+
 	/* ----- server setup ----- */
 
 	host?: string;

Index: sagemaker-code-editor/vscode/src/vs/server/node/serverServices.ts
===================================================================
--- sagemaker-code-editor.orig/vscode/src/vs/server/node/serverServices.ts
+++ sagemaker-code-editor/vscode/src/vs/server/node/serverServices.ts
@@ -12,7 +12,7 @@ import * as path from '..
 import { IURITransformer } from '../../base/common/uriIpc.js';
 import { getMachineId, getSqmMachineId, getdevDeviceId } from '../../base/node/id.js';
 import { Promises } from '../../base/node/pfs.js';
-import { ClientConnectionEvent, IMessagePassingProtocol, IPCServer, StaticRouter } from '../../base/parts/ipc/common/ipc.js';
+import { ClientConnectionEvent, IMessagePassingProtocol, IPCServer, ProxyChannel, StaticRouter } from '../../base/parts/ipc/common/ipc.js';
 import { ProtocolConstants } from '../../base/parts/ipc/common/ipc.net.js';
 import { IConfigurationService } from '../../platform/configuration/common/configuration.js';
 import { ConfigurationService } from '../../platform/configuration/common/configurationService.js';
@@ -255,6 +255,9 @@ export async function setupServerService
 		const channel = new ExtensionManagementChannel(extensionManagementService, (ctx: RemoteAgentConnectionContext) => getUriTransformer(ctx.remoteAuthority));
 		socketServer.registerChannel('extensions', channel);

+		const languagePackChannel = ProxyChannel.fromService<RemoteAgentConnectionContext>(accessor.get(ILanguagePackService), disposables);
+		socketServer.registerChannel('languagePacks', languagePackChannel);
+
 		// clean up extensions folder
 		remoteExtensionsScanner.whenExtensionsReady().then(() => extensionManagementService.cleanUp());
 

Index: sagemaker-code-editor/vscode/src/vs/server/node/webClientServer.ts
===================================================================
--- sagemaker-code-editor.orig/vscode/src/vs/server/node/webClientServer.ts
+++ sagemaker-code-editor/vscode/src/vs/server/node/webClientServer.ts
@@ -25,6 +25,7 @@ import { URI } from '..
 import { streamToBuffer } from '../../base/common/buffer.js';
 import { IProductConfiguration } from '../../base/common/product.js';
 import { isString, Mutable } from '../../base/common/types.js';
+import { getLocaleFromConfig, getBrowserNLSConfiguration } from '../../server/node/remoteLanguagePacks.js';
 import { CharCode } from '../../base/common/charCode.js';
 import { IExtensionManifest } from '../../platform/extensions/common/extensions.js';
 import { ICSSDevelopmentService } from '../../platform/cssDev/node/cssDevService.js';
@@ -245,7 +246,10 @@ export class WebClientServer {
 		};
 
 		// Prefix routes with basePath for clients
-		const basePath = getFirstHeader('x-forwarded-prefix') || this._basePath;
+		const proxyPath = this._environmentService.args["base-path"] || "/";
+		const base = relativeRoot(proxyPath);
+		const vscodeBase = relativePath(proxyPath);
+		const basePath = vscodeBase || getFirstHeader("x-forwarded-prefix") || this._basePath;
 
 		const queryConnectionToken = parsedUrl.query[connectionTokenQueryName];
 		if (typeof queryConnectionToken === 'string') {
@@ -287,7 +291,7 @@ export class WebClientServer {
 		let remoteAuthority = (
 			useTestResolver
 				? 'test+test'
-				: (getFirstHeader('x-original-host') || getFirstHeader('x-forwarded-host') || req.headers.host)
+				: (getFirstHeader('x-original-host') || getFirstHeader('x-forwarded-host') || req.headers.host || window.location.host)
 		);
 		if (!remoteAuthority) {
 			return serveError(req, res, 400, `Bad request.`);
@@ -333,6 +337,7 @@ export class WebClientServer {
 		} : undefined;
 
 		const productConfiguration: Partial<Mutable<IProductConfiguration>> = {
+			rootEndpoint: base,
 			embedderIdentifier: 'server-distro',
 			extensionsGallery: this._webExtensionResourceUrlTemplate && this._productService.extensionsGallery ? {
 				...this._productService.extensionsGallery,
@@ -364,7 +369,7 @@ export class WebClientServer {

 		const workbenchWebConfiguration = {
 			remoteAuthority,
-			serverBasePath: basePath,
+			serverBasePath: this._basePath,
			webviewEndpoint: staticRoute + '/out/vs/workbench/contrib/webview/browser/pre',
			userDataPath: this._environmentService.userDataPath,
 			_wrapWebWorkerExtHostInIframe,
@@ -385,22 +390,32 @@ export class WebClientServer {
 		};

 		const cookies = cookie.parse(req.headers.cookie || '');
-		const locale = cookies['vscode.nls.locale'] || req.headers['accept-language']?.split(',')[0]?.toLowerCase() || 'en';
+		const locale = this._environmentService.args.locale || await getLocaleFromConfig(this._environmentService.argvResource.fsPath) || cookies['vscode.nls.locale'] || req.headers['accept-language']?.split(',')[0]?.toLowerCase() || 'en';
 		let WORKBENCH_NLS_BASE_URL: string | undefined;
 		let WORKBENCH_NLS_URL: string;
 		if (!locale.startsWith('en') && this._productService.nlsCoreBaseUrl) {
 			WORKBENCH_NLS_BASE_URL = this._productService.nlsCoreBaseUrl;
 			WORKBENCH_NLS_URL = `${WORKBENCH_NLS_BASE_URL}${this._productService.commit}/${this._productService.version}/${locale}/nls.messages.js`;
 		} else {
-			WORKBENCH_NLS_URL = ''; // fallback will apply
+			try {
+				const nlsFile = await getBrowserNLSConfiguration(locale, this._environmentService.userDataPath);
+				WORKBENCH_NLS_URL = nlsFile
+					? `${vscodeBase}/vscode-remote-resource?path=${encodeURIComponent(nlsFile)}`
+					: '';
+			} catch (error) {
+				console.error("Failed to generate translations", error);
+				WORKBENCH_NLS_URL = '';
+			}
 		}

 		const values: { [key: string]: string } = {
			WORKBENCH_WEB_CONFIGURATION: asJSON(workbenchWebConfiguration),
 			WORKBENCH_AUTH_SESSION: authSessionInfo ? asJSON(authSessionInfo) : '',
 			WORKBENCH_WEB_BASE_URL: staticRoute,
 			WORKBENCH_NLS_URL,
-			WORKBENCH_NLS_FALLBACK_URL: `${staticRoute}/out/nls.messages.js`
+			WORKBENCH_NLS_FALLBACK_URL: `${staticRoute}/out/nls.messages.js`,
+			BASE: base,
+			VS_BASE: vscodeBase
 		};
 
 		// DEV ---------------------------------------------------------------------------------------
@@ -618,3 +633,60 @@ export class WebClientServer {
 		serveError(req, res, 500, error.message);
 	}
 }
+
+
+/**
+ * Remove extra slashes in a URL.
+ *
+ * This is meant to fill the job of `path.join` so you can concatenate paths and
+ * then normalize out any extra slashes.
+ *
+ * If you are using `path.join` you do not need this but note that `path` is for
+ * file system paths, not URLs.
+ */
+export const normalizeUrlPath = (url: string, keepTrailing = false): string => {
+	return url.replace(/\/\/+/g, "/").replace(/\/+$/, keepTrailing ? "/" : "")
+}
+
+/**
+ * Get the relative path that will get us to the root of the page. For each
+ * slash we need to go up a directory.  Will not have a trailing slash.
+ *
+ * For example:
+ *
+ * / => .
+ * /foo => .
+ * /foo/ => ./..
+ * /foo/bar => ./..
+ * /foo/bar/ => ./../..
+ *
+ * All paths must be relative in order to work behind a reverse proxy since we
+ * we do not know the base path.  Anything that needs to be absolute (for
+ * example cookies) must get the base path from the frontend.
+ *
+ * All relative paths must be prefixed with the relative root to ensure they
+ * work no matter the depth at which they happen to appear.
+ *
+ * For Express `req.originalUrl` should be used as they remove the base from the
+ * standard `url` property making it impossible to get the true depth.
+ */
+export const relativeRoot = (originalUrl: string): string => {
+	const depth = (originalUrl.split("?", 1)[0].match(/\//g) || []).length
+	return normalizeUrlPath("./" + (depth > 1 ? "../".repeat(depth - 1) : ""))
+}
+
+/**
+ * Get the relative path to the current resource.
+ *
+ * For example:
+ *
+ * / => .
+ * /foo => ./foo
+ * /foo/ => .
+ * /foo/bar => ./bar
+ * /foo/bar/ => .
+ */
+export const relativePath = (originalUrl: string): string => {
+	const parts = originalUrl.split("?", 1)[0].split("/")
+	return normalizeUrlPath("./" + parts[parts.length - 1])
+}
Index: sagemaker-code-editor/vscode/src/vs/workbench/contrib/extensions/browser/extensionsActions.ts
===================================================================
--- sagemaker-code-editor.orig/vscode/src/vs/workbench/contrib/extensions/browser/extensionsActions.ts
+++ sagemaker-code-editor/vscode/src/vs/workbench/contrib/extensions/browser/extensionsActions.ts
@@ -475,9 +475,6 @@ export class InstallAction extends Exten
 		if (this.extension.isBuiltin) {
 			return;
 		}
-		if (this.extensionsWorkbenchService.canSetLanguage(this.extension)) {
-			return;
-		}
 		if (this.extension.state !== ExtensionState.Uninstalled) {
			return;
		}
@@ -782,7 +779,7 @@ export abstract class InstallInOtherServ
 		}
 
 		if (isLanguagePackExtension(this.extension.local.manifest)) {
-			return true;
+			return false;
 		}
 
 		// Prefers to run on UI
@@ -2073,17 +2070,6 @@ export class SetLanguageAction extends E
 	update(): void {
 		this.enabled = false;
 		this.class = SetLanguageAction.DisabledClass;
-		if (!this.extension) {
-			return;
-		}
-		if (!this.extensionsWorkbenchService.canSetLanguage(this.extension)) {
-			return;
-		}
-		if (this.extension.gallery && language === getLocale(this.extension.gallery)) {
-			return;
-		}
-		this.enabled = true;
-		this.class = SetLanguageAction.EnabledClass;
 	}
 
 	override async run(): Promise<any> {
@@ -2100,7 +2086,6 @@ export class ClearLanguageAction extends
 	private static readonly DisabledClass = `${ClearLanguageAction.EnabledClass} disabled`;
 
 	constructor(
-		@IExtensionsWorkbenchService private readonly extensionsWorkbenchService: IExtensionsWorkbenchService,
 		@ILocaleService private readonly localeService: ILocaleService,
 	) {
 		super(ClearLanguageAction.ID, ClearLanguageAction.TITLE.value, ClearLanguageAction.DisabledClass, false);
@@ -2110,17 +2095,6 @@ export class ClearLanguageAction extends
 	update(): void {
 		this.enabled = false;
 		this.class = ClearLanguageAction.DisabledClass;
-		if (!this.extension) {
-			return;
-		}
-		if (!this.extensionsWorkbenchService.canSetLanguage(this.extension)) {
-			return;
-		}
-		if (this.extension.gallery && language !== getLocale(this.extension.gallery)) {
-			return;
-		}
-		this.enabled = true;
-		this.class = ClearLanguageAction.EnabledClass;
 	}
 
 	override async run(): Promise<any> {

Index: sagemaker-code-editor/vscode/src/vs/workbench/services/localization/electron-sandbox/localeService.ts
===================================================================
--- sagemaker-code-editor.orig/vscode/src/vs/workbench/services/localization/electron-sandbox/localeService.ts
+++ sagemaker-code-editor/vscode/src/vs/workbench/services/localization/electron-sandbox/localeService.ts
@@ -51,7 +51,8 @@ class NativeLocaleService implements ILo
 		@IProductService private readonly productService: IProductService
 	) { }
 
-	private async validateLocaleFile(): Promise<boolean> {
+	// Make public just so we do not have to patch all the unused code out.
+	public async validateLocaleFile(): Promise<boolean> {
 		try {
 			const content = await this.textFileService.read(this.environmentService.argvResource, { encoding: 'utf8' });
 
@@ -78,9 +79,6 @@ class NativeLocaleService implements ILo
 	}
 
 	private async writeLocaleValue(locale: string | undefined): Promise<boolean> {
-		if (!(await this.validateLocaleFile())) {
-			return false;
-		}
 		await this.jsonEditingService.write(this.environmentService.argvResource, [{ path: ['locale'], value: locale }], true);
 		return true;
 	}

Index: sagemaker-code-editor/vscode/build/gulpfile.reh.js
===================================================================
--- sagemaker-code-editor.orig/vscode/build/gulpfile.reh.js
+++ sagemaker-code-editor/vscode/build/gulpfile.reh.js
@@ -59,6 +59,7 @@ const serverResourceIncludes = [
 
 	// NLS
 	'out-build/nls.messages.json',
+	'out-build/nls.keys.json', // Required to generate translations.
 
 	// Process monitor
 	'out-build/vs/base/node/cpuUsage.sh',
Index: sagemaker-code-editor/vscode/src/vs/workbench/workbench.web.main.internal.ts
===================================================================
--- sagemaker-code-editor.orig/vscode/src/vs/workbench/workbench.web.main.internal.ts
+++ sagemaker-code-editor/vscode/src/vs/workbench/workbench.web.main.internal.ts
@@ -53,7 +53,7 @@ import './services/dialogs/browser/fileD
 import './services/host/browser/browserHostService.js';
 import './services/lifecycle/browser/lifecycleService.js';
 import './services/clipboard/browser/clipboardService.js';
-import './services/localization/browser/localeService.js';
+import './services/localization/electron-sandbox/localeService.js';
 import './services/path/browser/pathService.js';
 import './services/themes/browser/browserHostColorSchemeService.js';
 import './services/encryption/browser/encryptionService.js';
Index: sagemaker-code-editor/vscode/src/vs/workbench/contrib/localization/common/localizationsActions.ts
===================================================================
--- sagemaker-code-editor.orig/vscode/src/vs/workbench/contrib/localization/common/localizationsActions.ts
+++ sagemaker-code-editor/vscode/src/vs/workbench/contrib/localization/common/localizationsActions.ts
@@ -12,6 +12,7 @@ import { ServicesAccessor } from '../../
 import { ILanguagePackItem, ILanguagePackService } from '../../../../platform/languagePacks/common/languagePacks.js';
 import { ILocaleService } from '../../../services/localization/common/locale.js';
 import { IExtensionsWorkbenchService } from '../../extensions/common/extensions.js';
+import { language } from '../../../../base/common/platform.js';
 
 export class ConfigureDisplayLanguageAction extends Action2 {
 	public static readonly ID = 'workbench.action.configureLocale';
@@ -37,6 +38,16 @@ export class ConfigureDisplayLanguageAct
 
 		const installedLanguages = await languagePackService.getInstalledLanguages();
 
+		// Clean any existing (Current) text and add it back only to the correct locale
+		installedLanguages?.forEach(lang => {
+			if (lang.description) {
+				lang.description = lang.description.replace(/ \(Current\)$/, '');
+			}
+			if (lang.id?.toLowerCase() === language.toLowerCase()) {
+				lang.description = (lang.description || '') + localize('currentDisplayLanguage', " (Current)");
+			}
+		});
+
 		const disposables = new DisposableStore();
 		const qp = disposables.add(quickInputService.createQuickPick<ILanguagePackItem>({ useSeparators: true }));
 		qp.matchOnDescription = true;

Index: sagemaker-code-editor/vscode/src/vs/base/common/product.ts
===================================================================
--- sagemaker-code-editor.orig/vscode/src/vs/base/common/product.ts
+++ sagemaker-code-editor/vscode/src/vs/base/common/product.ts
@@ -56,6 +56,7 @@ export type ExtensionVirtualWorkspaceSup
 };
 
 export interface IProductConfiguration {
+	readonly rootEndpoint?: string;
 	readonly version: string;
 	readonly date?: string;
 	readonly quality?: string;