Index: sagemaker-code-editor/vscode/src/vs/server/node/remoteExtensionHostAgentServer.ts
===================================================================
--- sagemaker-code-editor.orig/vscode/src/vs/server/node/remoteExtensionHostAgentServer.ts
+++ sagemaker-code-editor/vscode/src/vs/server/node/remoteExtensionHostAgentServer.ts
@@ -124,13 +124,6 @@ class RemoteExtensionHostAgentServer ext
 			return void res.end(this._productService.commit || '');
 		}
 
-		// Delay shutdown
-		if (pathname === '/delay-shutdown') {
-			this._delayShutdown();
-			res.writeHead(200);
-			return void res.end('OK');
-		}
-
 		if (!httpRequestHasValidConnectionToken(this._connectionToken, req, parsedUrl)) {
 			// invalid connection token
 			return serveError(req, res, 403, `Forbidden.`);
@@ -630,18 +623,6 @@ class RemoteExtensionHostAgentServer ext
 		}
 	}
 
-	/**
-	 * If the server is in a shutdown timeout, cancel it and start over
-	 */
-	private _delayShutdown(): void {
-		if (this.shutdownTimer) {
-			console.log('Got delay-shutdown request while in shutdown timeout, delaying');
-			this._logService.info('Got delay-shutdown request while in shutdown timeout, delaying');
-			this._cancelShutdown();
-			this._waitThenShutdown();
-		}
-	}
-
 	private _cancelShutdown(): void {
 		if (this.shutdownTimer) {
 			console.log('Cancelling previous shutdown timeout');
